GTScript-Parallel-model.md

```diff
@@ -0,0 +1,110 @@
+The iteration domain is a 3d domain: `I` and `J` axes live on the horizontal spatial plane, and axis `K` represents the vertical spatial dimension.
+
+A GTScript `computation` is composed by one or more vertical `interval` specifications, each one of them representing a vertical loop over its vertical iteration range with a given iteration policy (`FORWARD`, `BACKWARD`, `PARALLEL`). Vertical `interval`s defined inside a computation should not have overlapping iteration ranges.
+
+Every statement inside a vertical interval is executed on the full horizontal `IJ` plane before the next one is executed. Individual statements of a computation are considered embarrassingly parallel in the horizontal and therefore there are no guarantees about the loop order in the `IJ` plane. Every iteration of the `K`-loop defined by a vertical interval can only start after the previous iteration have finished. Vertical intervals defined in a `computation` will be sorted according to the iteration policy and then executed in order, that is, every vertical interval will start its execution after all the previous vertical intervals in the iteration order have been fully executed. In the case of a `PARALLEL` iteration policy, vertical intervals will be ordered in any implementation-specific way (no guarantees).
+
+To summarize the model, we can say that:
+
+- *computations* are executed sequentially in the order they appear in the code,
+- vertical *intervals* are executed sequentially in the order defined by the *iteration policy* of the  *computation* (where `PARALLEL` means that order is not relevant and thus it will be implementation-specific),
```

Comment:
Is it enough to say that intervals in a `PARALLEL` computation can be reordered to allow us to actually schedule them independently? At least the "sequentially" on line 10 I would remove, after that I think it would be equivalent to not making any assumption on the order.

Reply:
Probably we should discuss the future of `PARALLEL` as it doesn't add extra semantic information to the parallel model, unless we will (also in the future) blindly apply `PARALLEL` instead of checking.

Currently `PARALLEL` doesn't imply anything: it doesn't have to be executed in parallel, it can be executed forward or backward as well.

Reply:
Just a comment: the future of PARALLEL is one of the issues I had collected for the workshop: https://hackmd.io/@egparedes/Bkf7qxibv